zonemaster/zonemaster-backend
# Backport fix user api

## Purpose

Secure the add_api_user method bit more and make it an optional feature.

## Context

Backport fix #839 for #838 to master

## Changes

* Use Plack::Middleware::ReverseProxy to set the environment according to the proxy headers
* Add two configuration keys API.enable_add_api_user (disabled by default) and API.enable_batch_jobs (enabled by default to minimize the change)

## How to test this PR

* Disabled the batch / user api.
* Running `./script/zmb add_api_user --username toto --api-key secret` should return the error `"Procedure 'add_api_user' not found"`.
* curl -s "Content-Type: application/json" --data '{"method":"add_batch_job","params":{"username":"toto","api_key":"toto", "domains":["example.com"]},"id":1,"jsonrpc":"2.0"}' 127.0.0.1:5000
* Enable the add_api_user method.
  * run the API behind a reverse proxy on the same host
    ```
    curl -s "Content-Type: application/json" --data '{"method":"add_api_user","params":{"username":"toto","api_key":"toto"},"id":1,"jsonrpc":"2.0"}' not.localhost/api  | jq
    ```
  * Try to add a user from a another host
  * If the query return `0` as result it works.
